src/views/validated-designs/server.ts

```diff
@@ -75,56 +80,176 @@ export function getHvdLandingProps(): ValidatedDesignsLandingProps | null {
+		const isHvdMetadata = pathParts.length === 4
+
+		const slug = `${product}-${category}`
 		if (isCategoryMetadata) {
 			const { title, description } = loadMetadata(
 				path.join(HVD_CONTENT_DIR, item)
 			)
-			categoryGroups[categorySlug] = {
-				slug: categorySlug,
+
+			hvdCategoryGroups.push({
+				slug,
+				title,
+				description,
 				product,
-				title: title,
-				description: description,
 				guides: [],
-			}
+			})
 		} else if (isHvdMetadata) {
 			const { title, description } = loadMetadata(
 				path.join(HVD_CONTENT_DIR, item)
 			)
-			const guideSlug = `/validated-designs/${pathParts.slice(0, -1).join('-')}`
-
-			// This assumes the property is already set, @TODO harden this
-			categoryGroups[categorySlug].guides = [
-				{
-					slug: guideSlug,
-					title,
-					description,
-					href: guideSlug,
-				},
-				...categoryGroups[categorySlug].guides,
-			]
+
+			// find the existing HvdCategoryGroup, because we traverse the files in order the HvdCategoryGroup YAML will always come before the HvdGuide YAML
+			const categoryGroup = hvdCategoryGroups.find(
+				(categoryGroup: HvdCategoryGroup) => {
+					return categoryGroup.slug === slug
+				}
+			)
+
+			const guideSlug = pathParts[2]
+			const categorySlug = `${slug}-${guideSlug}`
+
+			// build the category group
+			const pagesPath = path.join(product, category, guideSlug)
+			const pagesFiles = hvdRepoContents.filter((path: string) => {
+				return path.includes(pagesPath) && path.endsWith('.mdx')
+			})
+			const pages = pagesFiles.map((pagePath: string) => {
+				const pagePathParts = pagePath.split('/')
+				const pageFile = pagePathParts[pagePathParts.length - 1]
+
+				const slug = pageFile
+					.replace('.mdx', '')
+					.substring(pageFile.indexOf('-') + 1)
+
+				return {
+					slug,
+					title: slug.replaceAll('-', ' '),
+					filePath: path.join(HVD_CONTENT_DIR, pagePath),
+					href: `${basePath}/${categorySlug}/${slug}`,
+				}
+			})
+
+			categoryGroup.guides.push({
+				slug: categorySlug,
+				title,
+				description,
+				href: `${basePath}/${categorySlug}`,
+				pages,
+			})
 		}
 	})
 
+	return hvdCategoryGroups
+}
+
+export function getHvdCategoryGroupsPaths(): string[][] {
+	const categoryGroups = getHvdCategoryGroups()
+	// [[guide-slug], [guide-slug, page-slug]]
+	// e.g. [[terraform-operation-guide-adoption], [terraform-operation-guide-adoption, page-slug]]
+	const paths = []
+	for (const categoryGroup of categoryGroups) {
+		for (const guide of categoryGroup.guides) {
+			paths.push([guide.slug])
+			for (const page of guide.pages) {
+				paths.push([guide.slug, page.slug])
+			}
+		}
+	}
+
+	return paths
+}
+
+// @TODO: calculate the actual url from the headers
+function getMarkdownHeaders(markdown: string) {
+	const headerRegex = /^(#{1,6})\s*(.+)/gm
+	let match
+	const headers = []
+
+	while ((match = headerRegex.exec(markdown)) !== null) {
+		headers.push({
+			title: match[2].trim(),
+			url: '#test',
+		})
+	}
```

Comment:
I think using a `while` loop in production could open up the possibility errors. For Regex.exec(), MDN suggests replacing it with [String.prototype.matchAll()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/matchAll) to make it less error-prone. What do you think?

Reply:
Honestly this is a hack till I can figure out how the rest of our code extracts and adds data to headers in markdown. (Which is how we get internal links working on markdown headers.) Late yesterday I think I found a thread I wanted to chase down here: https://github.com/hashicorp/dev-portal/blob/b59f2dbe6b0da9dc1fb98b1bf126cb9182ac2155/src/lib/serialize-integration-markdown.ts#L28.